### PR TITLE
[WIP] upstream: dynamic cluster creation in request path

### PR DESCRIPTION
**Pull request by ramaraochavali**
*Description*:
Allows to create a static cluster dynamically on the request path based on request headers.

The following changes are proposed.

- Change the `ClusterManager's` `addOrUpdateCluster` method so that it can work when called from worker thread.
- Introduced a `dynamic_cluster` filter that uses the above api and creates a cluster based on following headers
        x-envoy-cluster : Defines the name of the cluster to be created.
        x-envoy-cluster-hosts: Host information in the form of host:port.

*Risk Level*: Low as this is an opt-in feature
*Testing*: Automated
*Docs Changes*: Will follow-up
*Release Notes*: Will follow-up

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@mattklein123 PR for the dynamic cluster creation we have been discussing. I thought it is better to iterate this over code so made some initial proposal. PTAL and we can iterate and work together for a final solution.

File: source/common/upstream/cluster_manager_impl.cc [hunk 1]
```diff
@@ -406,28 +407,35 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
...
-            }
-
-            auto thread_local_cluster = new ThreadLocalClusterManagerImpl::ClusterEntry(
-                cluster_manager, new_cluster, thread_aware_lb_factory);
-            cluster_manager.thread_local_clusters_[new_cluster->name()].reset(thread_local_cluster);
-            for (auto& cb : cluster_manager.update_callbacks_) {
-              cb->onClusterAddOrUpdate(*thread_local_cluster);
-            }
-          });
+  createOrUpdateThreadLocalClusterInternal(
```
Comment (ramaraochavali) [inline source/common/upstream/cluster_manager_impl.cc]: there will be a redundant message I think to the thread that this is called on initially. But considering that clusters are created rarely, that should be fine. But open to other options here.

File: source/extensions/filters/http/dynamic_cluster/dynamic_cluster.cc [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  const Http::LowerCaseString ClusterName{"x-envoy-cluster"};
+  const Http::LowerCaseString ClusterHosts{"x-envoy-cluster-hosts"};
+  const std::string cluster_name = headers.get(ClusterName)->value().c_str();
+  const std::string cluster_hosts = headers.get(ClusterHosts)->value().c_str();
+  std::istringstream ss(cluster_hosts);
+  std::string host, port;
+  std::getline(ss, host, ':');
+  std::getline(ss, port, ':');
+
+  // TODO: Make this part of filter config - possibly multiple templates for each type http1, http2
```
Comment (ramaraochavali) [inline source/extensions/filters/http/dynamic_cluster/dynamic_cluster.cc]: Couple of options here 
- Create Cluster object entirely based on http header information with sensible defaults to most of the fields.
- Provide filter config to select the cluster template and fill it with the fields from http headers.

**Comment by ramaraochavali** [comment]
@mattklein123 when you get time, could you PTAL at this approach? Based on your guidance, I will make further changes.

**Comment by mattklein123** [comment]
@ramaraochavali I will look this week.

**Comment by ramaraochavali** [comment]
@mattklein123 Thank you so much. Will wait for your feedback. If you think the dynamic_cluster filter is not generic enough or takes lot more thought to make it generic, I am OK adding lua wrapper to the api. I think the key part is the api.

File: source/common/upstream/cluster_manager_impl.cc [hunk 3]
```diff
@@ -619,6 +635,7 @@ Http::AsyncClient& ClusterManagerImpl::httpAsyncClientForCluster(const std::stri
 ClusterUpdateCallbacksHandlePtr
 ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks& cb) {
   ThreadLocalClusterManagerImpl& cluster_manager = tls_->getTyped<ThreadLocalClusterManagerImpl>();
+  cluster_manager.update_callbacks_.emplace_back(&cb);
```
Comment (ramaraochavali) [inline source/common/upstream/cluster_manager_impl.cc]: this is not related to this change. But I guess, cluster callbacks are missing. so added this.

File: source/extensions/filters/http/lua/lua_filter.cc [hunk 4]
```diff
@@ -200,6 +200,16 @@ int StreamHandleWrapper::luaHttpCall(lua_State* state) {
   }
 }
 
+int StreamHandleWrapper::luaAddorUpdateCluster(lua_State* state) {
```
Comment (ramaraochavali) [inline source/extensions/filters/http/lua/lua_filter.cc]: Added support of Lua. I think this may be more generic as people can pass whatever cluster definition they want in yaml here and the lua will just call `clustermanager`'s  `addOrUpdateCluster`.

**Comment by mattklein123** [comment]
@ramaraochavali I looked through and at a very high level I think this is interesting functionality that we should add to Envoy. The core cluster manager changes make sense to me (modulo comments that I have), as we move out from there it's less clear to me what should be built into Envoy directly at the filter/Lua level. That will need more discussion.

Here is how I suggest we proceed:
1) Let's close this PR and split it into multiple pieces.
2) First, let's do the core cluster manager changes that allow us to create clusters from worker threads, and wait for their creation to go to the main thread, and back to all worker threads. (I would so the implementation slightly differently from how you did it, but we can discuss that in the PR.).
3) Once the core code exists, we can open an issue to discuss design of add on components/filters/etc. that can use this code. I have lots of questions about the right programming model, reaping, limits on added clusters, etc. Lots to discuss.

Thank you!

**Comment by ramaraochavali** [comment]
@mattklein123 Thanks for your suggestion. I agree with splitting the PR in to multiple parts. I basically created this PR so that we get a feel of what we are talking about. I will open a fresh PR for api first. Regarding the add on components/filters, I do think that we can go incrementally there by exposing the basic cluster creation via Lua (as it seems more generic based on what I prototyped here) and expand it with other features like limits on added clusters, reaping, providing api to clean them if needed or automatically cleaning them etc. Will create an issue for that so that we can discuss.
